File: client/src-tauri/src/commands/expenses.rs
//! Expenses (gastos / caja chica) commands.

use secrecy::ExposeSecret;
use tauri::State;

use crate::http::{base, client, conn_error, error_message};
use crate::state::{SessionState, token_of};
use crate::types::Expense;

/// GET `/api/v1/expenses` (Bearer, cashier+). Optional `category` /
/// `payment_method` filters + `limit`. Returns the tenant's expenses
/// (egresos / caja chica). Requires cashier+ role — same ladder as the POS.
#[tauri::command]
pub async fn list_expenses(
    state: State<'_, SessionState>,
    server_url: String,
    category: Option<String>,
    payment_method: Option<String>,
    limit: Option<u32>,
) -> Result<Vec<Expense>, String> {
    let token = token_of(&state)?;
    let http = client();
    let base = base(&server_url);
    let mut req = http
        .get(format!("{base}/api/v1/expenses"))
        .bearer_auth(token.expose_secret());
    if let Some(c) = category.as_ref().filter(|s| !s.is_empty()) {
        req = req.query(&[("category", c)]);
    }
    if let Some(p) = payment_method.as_ref().filter(|s| !s.is_empty()) {
        req = req.query(&[("payment_method", p)]);
    }
    if let Some(n) = limit {
        req = req.query(&[("limit", n)]);
    }
    let resp = req.send().await.map_err(conn_error)?;
    if !resp.status().is_success() {
        return Err(error_message(resp).await);
    }
    resp.json()
        .await
        .map_err(|e| format!("Respuesta de gastos inválida del servidor: {e}"))
}

/// POST `/api/v1/expenses` (Bearer, cashier+). Body `NewExpense`: `category`,
/// `description`, `amount` (STRING, forwarded verbatim), optional
/// `payment_method` (defaults to `cash` server-side), `note`, and `incurred_at`
/// (RFC3339). Returns the created expense.
#[allow(clippy::too_many_arguments)]
#[tauri::command]
pub async fn create_expense(
    state: State<'_, SessionState>,
    server_url: String,
    category: String,
    description: String,
    amount: String,
    payment_method: Option<String>,
    note: Option<String>,
    incurred_at: Option<String>,
) -> Result<Expense, String> {
    let token = token_of(&state)?;
    let http = client();
    let base = base(&server_url);
    let mut body = serde_json::json!({
        "category": category,
        "description": description,
        "amount": amount,
    });
    if let Some(p) = payment_method.filter(|s| !s.is_empty()) {
        body["payment_method"] = serde_json::Value::String(p);
    }
    if let Some(n) = note.filter(|s| !s.is_empty()) {
        body["note"] = serde_json::Value::String(n);
    }
    if let Some(t) = incurred_at.filter(|s| !s.is_empty()) {
        body["incurred_at"] = serde_json::Value::String(t);
    }
    let resp = http
        .post(format!("{base}/api/v1/expenses"))
        .bearer_auth(token.expose_secret())
        .json(&body)
        .send()
        .await
        .map_err(conn_error)?;
    if !resp.status().is_success() {
        return Err(error_message(resp).await);
    }
    resp.json()
        .await
        .map_err(|e| format!("Respuesta de gasto inválida del servidor: {e}"))
}

File: client/src/api/expenses.ts
// Expenses (gastos / caja chica) wrappers (client/src-tauri/src/commands/expenses.rs).
import { invoke } from "@tauri-apps/api/core";

/** An expense / egreso (`ExpenseDto`). `amount` is a STRING (Decimal).
 *  `cash_session`/`supplier`/`note`/`created_by` are null when unset. */
export interface Expense {
  id: string;
  category: string;
  description: string;
  amount: string;
  payment_method: string; // "cash" | "bank" | "card" | "transfer"
  cash_session: string | null;
  supplier: string | null;
  note: string | null;
  created_by: string | null;
  incurred_at: string;
  created_at: string;
}

/** Fields the Gastos form sends to create an expense (`NewExpense`). `amount` is
 *  a STRING; `paymentMethod` defaults to `cash` server-side when omitted. */
export interface NewExpense {
  category: string;
  description: string;
  amount: string;
  paymentMethod?: string;
  note?: string;
  incurredAt?: string; // RFC3339
}

/** GET /api/v1/expenses (Bearer, cashier+). `category`/`paymentMethod`/`limit`
 *  optional filters. */
export function listExpenses(
  serverUrl: string,
  category?: string,
  paymentMethod?: string,
  limit?: number,
): Promise<Expense[]> {
  return invoke<Expense[]>("list_expenses", { serverUrl, category, paymentMethod, limit });
}

/** POST /api/v1/expenses (Bearer, cashier+) — record an expense. `amount` STRING. */
export function createExpense(serverUrl: string, e: NewExpense): Promise<Expense> {
  return invoke<Expense>("create_expense", {
    serverUrl,
    category: e.category,
    description: e.description,
    amount: e.amount,
    paymentMethod: e.paymentMethod,
    note: e.note,
    incurredAt: e.incurredAt,
  });
}
